pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -328,9 +346,26 @@ private SuccessResponse uploadSegmentInternal(FormDataMultiPart multiPart, Strin
             throw new ControllerApplicationException(LOGGER,
                 "Failed to get segment fetcher for download URI: " + downloadURI, Response.Status.BAD_REQUEST);
           }
-
-          // Download segment tar file to local
-          segmentFetcher.fetchSegmentToLocal(downloadURI, tempTarredSegmentFile);
+          if (segmentMetadata == null) {
+            // Download segment tar file to local
+            segmentFetcher.fetchSegmentToLocal(downloadURI, tempTarredSegmentFile);
```

Comment:
You need the segmentFetcher only if metadata is null, right? Can you push that code inside the if?